### PR TITLE
Configure Envoy TLS validation for istiod

### PR DESCRIPTION
**Pull request by lei-tang**
Please provide a description for what this PR is for: https://github.com/istio/istio/issues/19950

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[X ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: pilot/cmd/pilot-agent/main.go [hunk 1]
```diff
@@ -417,6 +419,15 @@ var (
 			// TODO: change Mixer and Pilot to use standard template and deprecate this custom bootstrap parser
 			if controlPlaneBootstrap {
 				if templateFile != "" && proxyConfig.CustomConfigFile == "" {
+					// CitadelCACertPath is the directory for Citadel CA certificate.
+					// TODO (lei-tang): reuse CitadelCACertPath in https://github.com/istio/istio/pull/20017
+					// after https://github.com/istio/istio/pull/20017 is merged.
+					CitadelCACertPath := "./etc/istio/citadel-ca-cert"
+					signCertAtKubernetesCA := true
+					if _, err := ioutil.ReadFile(CitadelCACertPath + "/" + bootstrap.CACertNamespaceConfigMapDataName); err == nil {
```
Comment (howardjohn) [inline pilot/cmd/pilot-agent/main.go]: nit: use path.join

**Review by howardjohn — APPROVED** [review]

File: tools/packaging/common/envoy_bootstrap_v2.json [hunk 2]
```diff
@@ -262,7 +262,11 @@
             ],
             "validation_context": {
               "trusted_ca": {
+                {{ if .sign_cert_at_kubernetes_ca }}
```
Comment (myidpt) [inline tools/packaging/common/envoy_bootstrap_v2.json]: I'm curious: istiod doesn't have a sidecar, right? Are you changing this for normal workloads?

Comment (lei-tang) [reply tools/packaging/common/envoy_bootstrap_v2.json]: Yes, this is for normal workloads.

File: tools/packaging/common/envoy_bootstrap_v2.json [hunk 3]
```diff
@@ -262,7 +262,11 @@
...
                 "filename": "./var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
+                {{ else }}
+                "filename": "./etc/istio/citadel-ca-cert/ca-cert-ns.pem"
```
Comment (howardjohn) [inline tools/packaging/common/envoy_bootstrap_v2.json]: what happens when a proxy tries to connect to a pilot that is configured differently? I am worried this is impossible to upgrade ?

**Review by howardjohn — CHANGES_REQUESTED** [review]

File: tools/packaging/common/envoy_bootstrap_v2.json (hunk 3 above)
Comment (lei-tang) [reply tools/packaging/common/envoy_bootstrap_v2.json]: The TLS configurations for proxy and pilot must be consistent.

Comment (howardjohn) [reply tools/packaging/common/envoy_bootstrap_v2.json]: how do you upgrade then? it sounds like it is impossible and we are back to the old "file mount to SDS" upgrade issue

Comment (lei-tang) [reply tools/packaging/common/envoy_bootstrap_v2.json]: When upgrade, the config for envoy and pilot will be upgraded to have consistent TLS configuration. The methodology in this PR is no different from existing implementation in this file that uses kubernetes.io/service account/ca.crt.

Comment (howardjohn) [reply tools/packaging/common/envoy_bootstrap_v2.json]: Its impossible to atomically update envoy's and Pilots. You will always have new pilot, then update the old envoys, so pilot needs to be backwards compatible with older envoy versions

**Review by howardjohn — APPROVED** [review]

**Pull request merged**